Address quoting issue impacting dev_paths containing @ symbols

jnhealy2 opened the pull request:
A dev_path containing an @ symbol will be parsed as a version in Spack and cause concretization errors.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

jnhealy2 committed:
Address quoting issue that casuses dev_paths containing @ symbols to parse as versions

psakievich commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2710,8 +2710,7 @@ def setup(
                 spack.spec.Spec(info["spec"]).constrained(
-                    "dev_path=%s"
-                    % spack.util.path.canonicalize_path(info["path"], default_wd=env.path)
+                    f'dev_path="{spack.util.path.canonicalize_path(info["path"], default_wd=env.path)}"'
```
Should we move quoting internal to `canonicalize_path`?

becker33 replied:
No, I think that adding it to `canonicalize_path` will cause problems everywhere that we use it *outside* of a spec constructor.

becker33 reviewed (changes requested):
Can you please add a regression test for this?

Otherwise LGTM.

jnhealy2 committed:
Fix additional location were dev_path was imporperly constructed

The dev_path must be quoted to avoid parsing issues when
paths contain '@' symbols. Also add tests to catch
regression of this behavior

jnhealy2 committed:
Merge branch 'develop' into dev-path-parsing-bug

psakievich commented:
@jnhealy2 looks like you just need to resolve a line length issue for the style check.  Spackbot can't handle that request so you'll have to do it manully

jnhealy2 committed:
Format to fix line length

jnhealy2 committed:
Merge branch 'dev-path-parsing-bug' of https://github.com/jnhealy2/spack into dev-path-parsing-bug

jnhealy2 committed:
fix failing tests

jnhealy2 committed:
Fix whitespace error

jnhealy2 committed:
Add binary_compatibility fixture to test

jnhealy2 committed:
Merge branch 'develop' into dev-path-parsing-bug

jnhealy2 committed:
Change string formatting to avoid multiline f string

psakievich commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -3236,3 +3238,24 @@ def test_spec_unification(unify, mutable_config, mock_packages):
+
+
+def test_concretize_dev_path_with_at_symbol_in_env(tmpdir, mock_packages, binary_compatibility):
+    spec_like = "trivial-smoke-test@1.0"
```
```suggestion
    spec_like = "develop-test@develop"
```

psakievich committed:
Update lib/spack/spack/test/concretization/core.py

psakievich committed:
Add tmate debug session

psakievich committed:
Merge remote-tracking branch 'origin/develop' into dev-path-parsing-bug

psakievich committed:
Revert "Add tmate debug session"

This reverts commit 24e2f77e3c7c319d551d71fb5fe6f6d7c73e370b.

psakievich committed:
Move test and refactor to use env methods

psakievich commented on lib/spack/spack/test/cmd/develop.py:
```diff
@@ -216,3 +216,23 @@ def test_develop_full_git_repo(
+
+def test_concretize_dev_path_with_at_symbol_in_env(mutable_mock_env_path, tmpdir, mock_packages):
+    spec_like = "develop-test@develop"
+    spec = spack.spec.Spec(spec_like)
```
```suggestion
```

psakievich committed:
Update lib/spack/spack/test/cmd/develop.py
